Finish revision of Runners, other v0.4 related changes.

VHellendoorn opened the pull request:
Breaking changes: updates a number of core classes to new revision. This PR contains a number of tangled commits, in brief:

- Code changes are primarily refactoring of the `LexerRunner` and `ModelRunner` to be dynamic objects, as per #3. Most of the changes are about propagating that difference to the other classes.
- `ModelRunner` now takes a `Model` and `LexerRunner` instance, and `LexerRunner` takes a `Lexer` and `Vocabulary` instance. This way, the runners connect the various stages of the pipeline, without those stages being aware of each other.
- This will break existing code, so both the CLI and the two examples have been updated to show the new usage.
- The `NestedModel` is not a separate model anymore, but instead an option to the `ModelRunner`. 
- `ModelRunner` does not have a `perline` option anymore, instead it refers to `LexerRunner` for this, which should hopefully be less ambiguous.
- `NGramModel` order is now no longer a global field in `ModelRunner`, but a settable parameter. A default global value is still set in `ModelRunner`, but this cannot be changed and is only used as default value.
- `InverseMixModel` is now the standard `MixModel` (see `standard()` method on the latter)
- This PR contains small changes to GigaCounter to hopefully address #5 (verification needed).

Hopefully these changes allow for more modular building of models, enabling more parallelism and making settings less opaque.
Going forward, I hope to fix the core interfaces as they are (so we can move to 1.x revisions) and I'll try to switch to a proper pull-based development model :)

dgopstein commented on src/main/java/slp/core/example/JavaRunner.java:
```diff
@@ -26,50 +26,62 @@ public static void main(String[] args) {
 		//         * Note, order matters! The most local model should be "right-most" so it is called upon last (i.e. "gets the final say")
 		//         * This is also why we apply the cache after the nested model
-		model = new InverseMixModel(model, new CacheModel());
+		model = MixModel.standard(model, new CacheModel());
```
I'm confused about how the `CacheModel` mixture will be used. It doesn't seem to be referenced by the `ModelRunner` or accessed directly?

VHellendoorn replied:
The cache model learns on the fly, meaning that it integrates every event only after being asked to model it. What facilitates this is that it is `dynamic` in nature; other models can get the same benefit by setting themselves to dynamic as well. The `AbstractModel` takes care of invoking the learning after modeling.

dgopstein replied:
As far as I can tell, though, the mix-model is being constructed after the model-runner is, and so the object containing the cache is never being invoked in this example.

VHellendoorn replied:
You're right, I see it now. I need to think about how to resolve this, since you shouldn't have to first compose the whole model before passing it to a `ModelRunner`. Maybe it should be a settable field, or not a field at all. I'll hold off on this until I see a more fitting solution.

dgopstein commented on src/main/java/slp/core/modeling/runners/ModelRunner.java:
```diff
@@ -0,0 +1,489 @@
+	
+	/**
+	 * <b>Important:</b> if self-testing is the goal, please enable it before enabling nested mode!
+	 * Otherwise, self-testing will learn a model over all test data now.
```
In general I'm wary of "important" notes in comments that aren't enforced in code. Is this the kind of constraint that could be integrated directly into the workflow? For example one way to do this might be by making "self-test" only accessible through the `.setNested` call. That way it'd be impossible to perform these operations in the wrong order.

VHellendoorn replied:
Yeah I really didn't like that part either; it was a pain just to get it to behave correctly regardless of the order in which those two (`setSelfTest` & `setNested`) are invoked. I was thinking of making a method called `configure`, in which both fields must be set at once, which sounds a bit like what you are suggesting. The downside is that such a method must ask for a `testRoot` File. Perhaps an overloaded version with and without that parameter would make sense?

dgopstein commented:
I think every design change here is an improvement. IMO this PR, or a close variant, should be committed. There are a couple minor issues I've pointed out specifically in the code, but nothing that's a blocker for me.

I have slight concerns about integrating the `NestedModel` into the `ModelRunner` in that it's slightly less customizable/configurable now, however that may not actually be a problem in practice.

This fix does in fact fix issue #5 for me. Thank you!

VHellendoorn commented:
I revised the update a bit to integrate the comments above and some off-line ones as well. The following new changes are added, in order of importance:
- I simplified the Lexer and LexerRunner interface, and enforce per-line setting in its constructor, hopefully reducing some confusion about that setting. 
- The NestedModel is re-instated as a separate model. It's not quite perfect yet, but it will do for now
- The ModelRunner's statistics now always assume a start-of-sentence marker, whether added by the LexerRunner or present in the data, and skips the first token of every file/sentence in collecting statistics.
- I updated documentation and examples throughout (and fixed the latter)
- I changed CacheModel's dynamic nature slightly to be more accurate for updating
- I removed some of ModelRunner's less useful methods and simplified its interface.

This seems more in line with what I was hoping for in the update, thanks for all the feedback. Let me know if there are any overall issues to address, otherwise I'm moving it forward.

dgopstein commented:
This revision's interface was very natural to interact with. Looks great to me!

VHellendoorn commented:
Excellent. I'm tagging on some minor convenience methods for and merging
